llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp

```diff
@@ -363,8 +363,28 @@ static Value *foldLogOpOfMaskedICmps_NotAllZeros_BMask_Mixed(
   // deduce anything from it.
   // For example,
   // (icmp ne (A & 12), 0) & (icmp eq (A & 3), 1) -> no folding.
-  if ((*BCst & *DCst) == 0)
+  if ((*BCst & *DCst) == 0) {
+    // Try to fold isNaN idiom:
+    // (icmp ne (A & FractionBits), 0) & (icmp eq (A & ExpBits), ExpBits) ->
+    // isNaN(A)
+    Value *Src;
+    if (match(A, m_ElementWiseBitCast(m_Value(Src))) &&
+        Src->getType()->getScalarType()->isIEEELikeFPTy() &&
+        !Builder.GetInsertBlock()->getParent()->hasFnAttribute(
```

Comment:
Given that we are coming from explicit float, do we need the `noimplicitfloat` check? IIUC `noimplicitfloat` would only be preventing us from converting the raw integer case to float.